tensorboard/data/server/logdir.rs

```diff
@@ -15,68 +15,48 @@ limitations under the License.
 
 //! Loader for many runs under a directory.
 
-use log::{error, warn};
-use std::collections::{HashMap, HashSet};
+use log::warn;
+use std::collections::HashMap;
+use std::io::{self, Read};
 use std::path::{Path, PathBuf};
-use walkdir::WalkDir;
 
 use crate::commit::Commit;
 use crate::run::RunLoader;
 use crate::types::Run;
 
-/// A loader for a log directory, connecting a filesystem to a [`Commit`] via [`RunLoader`]s.
-pub struct LogdirLoader<'a> {
-    commit: &'a Commit,
-    logdir: PathBuf,
-    runs: HashMap<Run, RunState>,
-    checksum: bool,
-}
+/// A TensorBoard log directory, with event files organized into runs.
+pub trait Logdir {
+    /// Type of output stream for reading event files under this log directory.
+    type File: Read;
 
-struct RunState {
-    /// Stateful loader for this run.
-    loader: RunLoader,
-    /// Path to this run's directory relative to the root logdir.
-    relpath: PathBuf,
-    /// Logdir-relative paths to other directories that normalize to the same name as this run
-    /// after lossy Unicode conversion.
-    ///
-    /// Directories on the filesystem need not be valid Unicode. On Unix, for instance, they're
-    /// arbitrary byte sequences without `\0` or `/`. But TensorBoard wants run names to be strings
-    /// for human display. Thus, we apply [`OsStr::to_string_lossy`] to convert run directory paths
-    /// to run names. As the name suggests, this is a lossy conversion: it replaces non-Unicode
-    /// sequences with U+FFFD. This creates an edge case wherein two different directories may map
-    /// to the same run name. In such a case, we merge the two directories into one run, and print
-    /// a warning.
+    /// Finds all event files under the log directory.
     ///
-    /// This set tracks the directories other than `self.relpath` that comprise this run. It is
-    /// non-empty only when we hit this edge case: i.e., when there are multiple filesystem paths
-    /// that are invalid Unicode and collide. It's used for deduplicating warnings, so that we
-    /// don't warn about the directory again on every load cycle.
+    /// Event files within each run should be emitted in chronological order. Canonically, a file
+    /// is an event file if its basename contains [`EVENT_FILE_BASENAME_INFIX`] as a substring.
+    fn discover(&self) -> io::Result<HashMap<Run, Vec<PathBuf>>>;
```

Comment:
Probably fine for now, but may eventually want this to be an iterator so that we don't necessarily need to block postprocessing of the run discovery results on a full scan (since those can be quite expensive).

Reply:
Indeed. I was also wondering whether it might be a good fit to go full
async here:

```rust
#[async_trait::async_trait]
pub async trait Logdir {
    /// A stream of discovered event files and associated runs.
    type Discovery: Stream<Item = Result<(Run, PathBuf)>>;
    /// Type of output stream for reading event files under this logdir.
    type File: tokio::AsyncRead;

    /// Finds event files... they should be grouped by run... etc.
    async fn discover(&self) -> Self::Discovery;
    /// Opens an event file... it can be read asynchronously... etc.
    async fn open(&self, path: &Path) -> io::Result<Self::File>;
}
```

That way, with just 8 reload threads, you could have *n* runs reloading
at any given time, with *n* − 8 of them waiting on network. One downside
is that this approach gives us some nice implicit throttling to ensure
that we’re not hammering the user’s network stack or GCS account. I’m
also not certain about the lifetimes—I suspect that we can make it work
out without needing `unsafe`, but I’m not certain.

So yeah, agreed, and let’s keep it as is for now.

> (since those can be quite expensive)

For reference, I can list 33K objects in about 3 seconds. So yes, slow,
but imho manageable.